numba/core/typing/builtins.py

```diff
@@ -825,7 +825,10 @@ def typer(val):
                 sig = fnty.get_call_type(self.context, (val, types.DType(ty)),
                                          {})
                 return sig.return_type
-            elif isinstance(val, (types.Number, types.Boolean, types.IntEnumMember)):
+            elif isinstance(val, (types.Number,
+                                  types.Boolean,
+                                  types.IntEnumMember,
+                                  types.Optional)):
```

Comment:
Does this permit something that is optional but not like a number?

Reply:
In current mainline, we have this, so I think we allow casting optional to any type in the lowering level?
https://github.com/numba/numba/blob/d70830f2262c2a911a3eb6e81781fe070a2e8891/numba/core/optional.py#L112

Reply:
This PR will let typeinfer also allow casting optional to any type in the typing level. but I could be wrong in some unknown aspect..